api/environments/sdk/services.py

```diff
@@ -0,0 +1,117 @@
+
+IdentityAndTraits: TypeAlias = tuple[Identity, list[Trait]]
+
+
+def get_transient_identity_and_traits(
+    environment: Environment,
+    sdk_trait_data: list[SDKTraitData],
+) -> IdentityAndTraits:
+    """
+    Get a transient `Identity` instance with a randomly generated identifier.
+    All traits are marked as transient.
+    """
+    return (
+        (
+            identity := _get_transient_identity(
+                environment=environment,
+                identifier=get_transient_identifier(sdk_trait_data),
+            )
+        ),
+        identity.generate_traits(_ensure_transient(sdk_trait_data), persist=False),
+    )
+
+
+def get_identified_transient_identity_and_traits(
+    environment: Environment,
+    identifier: str,
+    sdk_trait_data: list[SDKTraitData],
+) -> IdentityAndTraits:
+    """
+    Get a transient `Identity` instance.
+    If present in storage, it's a previously persisted identity with its traits,
+    combined with incoming traits provided to `sdk_trait_data` argument.
+    All traits constructed from `sdk_trait_data` are marked as transient.
+    """
+    sdk_trait_data = _ensure_transient(sdk_trait_data)
+    if identity := Identity.objects.filter(
+        environment=environment,
+        identifier=identifier,
+    ).first():
+        return identity, identity.update_traits(sdk_trait_data)
+    return (
+        identity := _get_transient_identity(
+            environment=environment,
+            identifier=identifier,
+        )
+    ), identity.generate_traits(sdk_trait_data, persist=False)
+
+
+def get_persisted_identity_and_traits(
+    environment: Environment,
+    identifier: str,
+    sdk_trait_data: list[SDKTraitData],
+) -> IdentityAndTraits:
+    """
+    Retrieve a previously persisted `Identity` instance or persist a new one.
+    Traits are persisted based on the organisation-level setting or a
+    `"transient"` attribute provided with each individual trait.
+    """
+    identity, created = Identity.objects.get_or_create(
+        environment=environment,
+        identifier=identifier,
+    )
+    persist_trait_data = environment.project.organisation.persist_trait_data
+    if created:
+        return identity, identity.generate_traits(
+            sdk_trait_data,
+            persist=persist_trait_data,
+        )
+    if persist_trait_data:
+        return identity, identity.update_traits(sdk_trait_data)
+    return identity, list(
+        {
+            trait.trait_key: trait
+            for trait in chain(
+                identity.identity_traits.all(),
+                identity.generate_traits(sdk_trait_data, persist=False),
+            )
+        }.values()
+    )
+
+
+def get_transient_identifier(sdk_trait_data: list[SDKTraitData]) -> str:
+    return hashlib.sha256(
+        "".join(
+            f'{trait["trait_key"]}{trait["trait_value"]["value"]}'
+            for trait in sorted(sdk_trait_data, key=itemgetter("trait_key"))
+        ).encode(),
```

Comment:
This is a good approach but I'm not sure it works how we want it to if no traits are provided? I think that would then mean that the same identifier is used every time. Maybe that's what we want, but I'm not sure it is?

Reply:
The concern is valid. I say let's fall back to uuid4 in case traits are empty.

Reply:
Agreed.

Reply:
Isn't that just a call to `/flags`?

Reply:
No, I don't think so. Think about percentage split / MV flags.

Reply:
> Isn't that just a call to /flags?

Essentially, yes, but 

> We should definitely not allow that?

I don't think we _definitely_ should. The identifier is being returned now, so people are able to make use of it in the future.

Reply:
The fallback is now implemented both here and in the Edge API PR.

Reply:
I'm happy with this. @gagantrivedi I'm keen to know your thoughts too?

Reply:
I think I'd prefer mv evaluation to be stable, i.e: if the identifier doesn't exists maybe use a hash of empty string or environment key or something

Reply:
@gagantrivedi When designing this, I've decided that the semantics for MV evaluation would be the following:

1. If you want a guaranteed stable MV evaluation, provide an identifier.
2. If you don't provide an identifier, you're not entitled to stable MV evaluation. If you don't offer an identifier, the API will generate one for you that you'll be able to reuse, along with the MV bucket that it's being put into.

(There's a [Slack thread](https://flagsmith.slack.com/archives/CTF0THS2D/p1722616527027659) with more context)

That being said, IMO, using a trait hash to stabilise MV for null identifiers is a good idea. However, placing all of the identifier-less + trait-less transient identities in the same bucket is definitely not desirable given the semantics listed above, as most of these requests are going to serve distinct anonymous users.

Reply:
Sounds good!